s2and/model.py

```diff
@@ -664,9 +664,35 @@ def predict_incremental(self, block_signatures: List[str], dataset: ANDData):
                     best_dist = average_dist
             if best_cluster_id is not None:
                 # undo the reclustering step
+                constraint_found = False
                 if best_cluster_id in recluster_map:
                     best_cluster_id = recluster_map[best_cluster_id]  # type: ignore
-                pred_clusters[f"{best_cluster_id}"].append(unassigned_signature)
+
+                    # restrict reclusterings that would add a new name incompatibility to the main cluster
+                    main_cluster_signatures = cluster_seeds_require_inverse[best_cluster_id]
+                    all_firsts = set(
+                        [
+                            dataset.signatures[signature_id].author_info_first_normalized_without_apostrophe
+                            for signature_id in main_cluster_signatures
+                        ]
+                    )
+                    for main_cluster_signature in main_cluster_signatures:
+                        first_1 = dataset.signatures[
```

Comment:
Change name to `first_assigned`

Reply:
Actually, it seems like we can get a set of all the `first_assigned` above, right? Instead of having this inner for loop happen repeatedly.

Reply:
do you just mean to loop over the `all_firsts` set?

Reply:
instead of `main_cluster_signatures`

Reply:
Yeah. It just avoids dict lookups, but those are cheap. Either way.